src/cpp/session/SessionRUtil.cpp

```diff
@@ -0,0 +1,161 @@
+/*
+ * SessionRUtil.cpp
+ *
+ * Copyright (C) 2009-2015 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <session/SessionRUtil.hpp>
+
+#include <core/Error.hpp>
+#include <core/Log.hpp>
+#include <core/Exec.hpp>
+#include <core/Macros.hpp>
+
+#include <core/FileSerializer.hpp>
+
+#include <r/RExec.hpp>
+
+#include <session/SessionModuleContext.hpp>
+
+#include <boost/regex.hpp>
+
+#include "modules/shiny/SessionShiny.hpp"
+
+namespace rstudio {
+
+using namespace core;
+
+namespace session {
+namespace r_utils {
+
+namespace {
+
+boost::regex& reYaml()
+{
+   static boost::regex instance("^[\\s\\n]*---\\s*(.*?)---\\s*(?:$|\\n)");
+   return instance;
+}
+
+} // anonymous namespace
+
+bool hasYamlHeader(const FilePath& filePath)
```

Comment:
rmarkdown has a function to this which we could also call (that would keep us consistent, as it currently stands I think your regex is more liberal in what it considers front matter). The name of the function is partition_yaml_front_matter and it's here: https://github.com/rstudio/rmarkdown/blob/7cf3037663c8df05a82bd270cda52f5f702fd0af/R/output_format.R#L509

@jmcphers Also does some YAML handling on the server and perhaps one of these routines could be used.